smst_testutil.go

```diff
@@ -0,0 +1,81 @@
+package smt
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+)
+
+// SMSTWithStorage Wraps an SMST with a mapping of value hashes to values with sums (preimages), for use in tests.
+// Note: this doesn't delete from preimages, since there could be duplicate stored values.
+type SMSTWithStorage struct {
+	*SMST
+	preimages MapStore
+}
+
+func (smst *SMSTWithStorage) Update(key, value []byte, sum uint64) error {
+	err := smst.SMST.Update(key, value, sum)
+	if err != nil {
+		return err
+	}
+	valueHash := smst.digestValue(value)
+	var sumBz [sumSize]byte
+	binary.BigEndian.PutUint64(sumBz[:], sum)
+	value = append(value, sumBz[:]...)
+	err = smst.preimages.Set(valueHash, value)
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+func (smst *SMSTWithStorage) Delete(key []byte) error {
+	err := smst.SMST.Delete(key)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Get gets the value and sum of a key from the tree.
+func (smst *SMSTWithStorage) GetValueSum(key []byte) ([]byte, uint64, error) {
+	valueHash, sum, err := smst.Get(key)
+	if err != nil {
+		return nil, 0, err
+	}
+	value, err := smst.preimages.Get(valueHash)
+	if err != nil {
+		var invalidKeyError *InvalidKeyError
+		if errors.As(err, &invalidKeyError) {
+			// If key isn't found, return default value and sum
+			return defaultValue, 0, nil
+		} else {
+			// Otherwise percolate up any other error
+			return nil, 0, err
+		}
+	}
+	var sumBz [sumSize]byte
+	copy(sumBz[:], value[len(value)-sumSize:])
+	storedSum := binary.BigEndian.Uint64(sumBz[:])
+	if storedSum != sum {
+		return nil, 0, fmt.Errorf("sum mismatch for %s: got %d, expected %d", string(key), storedSum, sum)
+	}
+	return value[:len(value)-sumSize], storedSum, nil
+}
+
+// Has returns true if the value at the given key is non-default, false
+// otherwise.
+func (smst *SMSTWithStorage) Has(key []byte) (bool, error) {
+	val, sum, err := smst.GetValueSum(key)
+	return !bytes.Equal(defaultValue, val) || sum != 0, err
```

Comment:
!(defaultValue && sum == 0) == !defaultValue || sum != 0 (de morgans law)

Reply:
@Olshansk Looking into this more from this example we have the following current functionality:

```go
package main

import (
	"crypto/sha256"
	"fmt"

	"github.com/pokt-network/smt"
)

func main() {
	nodeStore := smt.NewSimpleMap()
	tree := smt.NewSparseMerkleSumTree(nodeStore, sha256.New())
	tree.Update([]byte("key"), []byte(nil), 5)
	tree.Update([]byte("key2"), []byte("val2"), 5)
	tree.Update([]byte("key3"), []byte("val3"), 5)
	tree.Update([]byte("key4"), []byte("val4"), 5)
	root := tree.Root()
	proof, err := tree.Prove([]byte("key"))
	if err != nil {
		panic(err)
	}
	if valid := smt.VerifySumProof(proof, root, []byte("key"), nil, 5, tree.Spec()); valid { // valid
		fmt.Println("valid")
	} else {
		fmt.Println("invalid")
	}

	nodeStore2 := smt.NewSimpleMap()
	tree2 := smt.NewSparseMerkleTree(nodeStore2, sha256.New())
	tree2.Update([]byte("key"), []byte(nil))
	tree2.Update([]byte("key2"), []byte("val2"))
	tree2.Update([]byte("key3"), []byte("val3"))
	tree2.Update([]byte("key4"), []byte("val4"))
	root2 := tree2.Root()
	proof2, err := tree2.Prove([]byte("key"))
	if err != nil {
		panic(err)
	}
	if valid := smt.VerifyProof(proof2, root2, []byte("key"), nil, tree2.Spec()); valid { // invalid
		fmt.Println("valid")
	} else {
		fmt.Println("invalid")
	}
}
```

This makes me think that although we can set a `nil` value in both trees it will still be a placeholder. For the SMST this is true **ONLY** if we do not give it a sum. This aligns with the check in the `Has` function as if either the sum or the value are set the key is in the tree and is not a placeholder

Reply:
` if either the sum or the value are set the key is in the tree and is not a placeholder`

I feel like this implies that we can set a value to be empty (i.e. the placeholder) with a non-zero weight/sum. I can potentially see arguments for it, but feel like it creates room for error/confusion. 

In your example above, what I'm seeing is:
1. Empty key with weight -> valid proof
2. Empty key w/o weight -> invalid proof
3. Empty key with weight -> has=True
4. Empty key w/o weight -> has=False

I'm open to be swayed in the other direction, but I think that empty (i.e. placeholder) leaves should not be able to have a weight.

Reply:
I think we should allow ppl to set `nil` keys as they exist anyway, setting a key to `nil` is basically a noop, it does change the tree but the key is still a placeholder and this doesn't "count" as being in the tree. When we add a weight suddenly it does count as being in the tree as its not the default value anymore.

With this in mind I think the current functionality is correct. Weird edge case that nobody will probably utilise but I think the logic is sound. If the key is default in **every** then its not in the tree (placeholder) but if it has a value and 0 sum - in the tree, if it has no value but has a sum - in the tree. Think about the reverse should a key-value pair with a 0 sum be a placeholder?

Reply:
Actually all of these will change the root from the default placeholder `[32]byte` which may be confusing. I think this is because they are all actually creating a new leaf node thus changing the tree. However depending on the leaf's value its either the same as a placeholder leaf or not and thus doesn't count as being included in the tree.